### PR TITLE
CFE-4081: Fixed commiting of changes made by set-input

### PR DESCRIPTION
**Pull request by olehermanse**
More work remains to fix committing in `cfbs add`.

### CODE REVIEW

**Review by larsewi — APPROVED** [review]
ACK

File: tests/shell/032_set_input_unordered.sh [hunk 1]
```diff
@@ -40,3 +40,12 @@ echo '[
...
 ]' | cfbs --log=debug set-input delete-files -
+
+# Error if the file has never been added:
+git ls-files --error-unmatch delete-files/input.json
+
+# Error if there are staged (added, not yet commited)
+git diff --exit-code --staged
+
+# Error if there are uncommited changes (to tracked files):
+git diff --exit-code
```
Comment (vpodzime) [inline tests/shell/032_set_input_unordered.sh:51]: Looks like these checks should be in a nicely-named function in some shared bash source.

Comment (olehermanse) [reply tests/shell/032_set_input_unordered.sh:51]: yep, good idea

File: cfbs/commands.py [hunk 2]
```diff
@@ -1067,9 +1068,21 @@ def _compare_list(a, b):
...
 
     path = os.path.join(name, "input.json")
-    log.debug("Writing json to file '%s'" % path)
-    write_json(path, data)
 
+    log.debug("Comparing with data already in file '%s'" % path)
+    old_data = read_json(path)
+    if old_data == data:
+        log.debug("Input data for '%s' unchanged, nothing to write / commit" % name)
+        return 0
```
Comment (vpodzime) [inline cfbs/commands.py:1076]: This should, ideally, be done before the more expensive checks/comparison above.

File: cfbs/commands.py [hunk 3]
```diff
@@ -1067,9 +1068,21 @@ def _compare_list(a, b):
...
+
+    log.debug("Input data for '%s' changed, writing json to file '%s'" % (name, path))
+    write_json(path, data)
+    if config.get("git", False):
+        git_commit(
+            "Set the input for module '%s'" % name,
+            edit_commit_msg=False,
```
Comment (vpodzime) [inline cfbs/commands.py:1083]: No interactivity here and prompting the user if they want to modify the commit message?

Comment (olehermanse) [reply cfbs/commands.py:1083]: Right, I am submitting some changes which should explain / clear this up.

File: cfbs/commands.py (hunk 2 above)
Comment (olehermanse) [reply cfbs/commands.py:1076]: I was thinking it makes sense to always do the expensive checks (validation), even if the (potentially bad) data is already there.

File: cfbs/commands.py (hunk 3 above)
Comment (olehermanse) [reply cfbs/commands.py:1083]: WIP: https://github.com/cfengine/cfbs/pull/145

**Pull request merged**